### PR TITLE
Updating readme docs.

### PR DESCRIPTION
**Pull request by thenewt15**
This is a re-write of several sections in the readme. Added a lot of stuff I wish I knew when I was getting started with the library. 

The diff might be hard to sort through because there are so many changes...easiest way to read it may be to copy the markdown into a viewer like [dillinger.io](https://dillinger.io/).

- Created a "Quickstart" section. Tried to keep it super simple and discuss more advanced customization after that section. The Quickstart section uses `persistReducer` instead of `persistCombineReducers`. Since most people will come into here already having redux set up (with combineReducers already set up), I thought perhaps we could lead people toward `persistCombineReducers` when discussing more advanced customizations. That way we aren't asking them to refactor combineReducers in the QuickStart.
- I elaborated a little more on nested persistors and think this will help people take better advantage of this feature.
- Tweaked the the section on transforms and included a new example.
- I moved all the details of v4 to v5 migration to a separate file. I've seen a lot of other libraries do this where they have a standalone Migration Guide. I didn't edit the contents of this migration guide. Just a copy/paste from stuff in the current readme. 
- I didn't get around to editing the section on migrations.

### CODE REVIEW

**Comment by rt2zz** [comment]
thanks for doing this, definitely going to take a lot of these changes. The one that I am uncertain about is recommending persistReducer over persistCombineReducers. The main difference therein is the merge strategy, persistCombineReducers applies two levels of shallow merge whereas persistReducer only does 1 level shallow merge. Need to think on this a bit longer..

**Comment by thenewt15** [comment]
Would using persistReducer and with this config result in the exact same thing as using persistCombineReducers?

```
const persistConfig = {
  key: 'root',
  storage: storage,
  stateReconciler: autoMergeLevel2
}
```

Is there a benefit to using persistCombineReducers other than its merge level?

I'm trying to sort out why persistCombineReducers exists at all. It seems to do two things:
1. Defaults stateReconciler to autoMergeLevel2, which you could set explicitly with persistReducer.
2. Calls combineReducers for you. Is there a benefit to this though? People will likely already have combineReducers set up when they start implementing this library. What do they gain by passing all their reducers into persistCombineReducers rather than creating this root reducer themselves with combineReducers?

**Comment by thenewt15** [comment]
@rt2zz a follow-up question for you...

I'm trying to finalize my data store on my current project and ultimately just trying to figure out whether I should be using autoMergeLevel2 for my top-level PersistReducer (i.e. the PersistReducer that wraps my root reducer returned from combineReducers).

I noticed that the persistCombineReducers function does this for you (uses autoMergeLevel2), but I can't wrap my head around why this is needed. autoMergeLevel1 looks like it makes a copy of state, which should successfully notify all my container components via mapStateToProps. What additional benefit would I get by having copies made two levels deep?

**Comment by rt2zz** [comment]
@thenewt15 correct, persistCombineReducers is really no more than defaulting to autoMergeLevel2 and applying combineReducers. Those can be done separately if desired. We put them together in one method actually in an attempt to reduce confusion, since "autoMergeLevelX" is not a very user friendly name. Generally speaking the intent of autoMergeLevel2 is to work at the top level, whereas autoMergeLevel1 is intended for use when nested within your reducer tree.

The only difference between the two reconcilers is how many levels of state they shallow merge. If you want initialState merged in with incoming rehydrated state then you need level 2, otherwise level 1 is perfect.

Sorry for confusion and slow response. I know right now stateReconciler is the biggest pain point around understanding and implementing redux persist. I am looking for ways to improve the api around this!

tldr; yes using persistReducer (which is default autoMergeLevel1) is quite fine, just be aware that your sub reducer initial state's will not be shallow merged in during rehydrate.

**Comment by thenewt15** [comment]
@rt2zz I can add something to my PR on how merging works. I'd like to clarify a couple things about it though. Are these statements true?

1. Once the persisted state is retrieved from storage on app launch, redux-persist dispatches a REHYDRATE action. After the REHYDRATE action is sent to all your reducers, the merging process begins. 
2. The term "rehydrated state" refers to the situation where you catch the REHYDRATE action in a reducer and update the state object. The state returned from this REHYDRATE case is the "rehydrated state."
3. The merging process checks whether you changed the state object when the REHYDRATE action was dispatched (i.e. whether the rehydrated state is equal to the initial redux state). If it was changed, then the initial redux state is updated with this rehydrated state. If it was not changed, the initial redux state is updated with whatever was persisted (i.e. inboundState). 
4. redux-persist v5 introduced auto-merging, which means you no longer need to catch the REHYDRATE action within your reducers. Without catching this action, the initial redux state will simply be updated with any persisted state.


Ok, so regarding the two merge levels, I'm still confused about when you'd want to use autoMergeLevel2. Let's say I have a state object like this.
```
{
  auth: { currentUser: { firstName: 'Sally' }, isLoggingIn: false },
  otherStateKey: { ... }
}
```
With autoMergeLevel1, I've noticed that my restored redux state will include any changes I make to `isLoggingIn`, or even to `currentUser.firstName`, if I update these from the REHYDRATE action in my Auth Reducer.
```
  case (REHYDRATE): 
    return {
      ...state,
      isLoggingIn: true,
      currentUser: { ...state.currentUser, firstName: 'Bob' }
    };
```
Because I'm always returning a copy of state (like you always should), it doesn't matter how deep or shallow these changes go. I actually haven't been able to find a case where autoMergeLevel1 didn't pick up a change I made to state. Can you provide an example of a state change that wouldn't be detected?

**Comment by levsero** [comment]
@thenewt15 how would it work if you don't want to manually listen for `case REHYDRATE` in all your reducers? I assume this is what the deep merge is for, because the rehydrated state is not going to `===` the default state, so you need to deep merge the keys to merge them.

**Comment by thenewt15** [comment]
thank you @levsero, I misunderstood what the merge process was trying to handle. After looking at the code, I'm seeing that its meant to handle changes to your initial redux state. So if your auth state had its initial state set like this.
```
const INITIAL_STATE = {
  currentUser: null,
  isLoggingIn: false,
};
```
And then you released a new version of your app that set it like this.
```
const INITIAL_STATE = {
  currentUser: null,
  isLoggingIn: false,
  error: ''
};
```
You'd obviously want your reconciled state object to include this new error key. But `autoMergeLevel1` would cause it to not be included, because it would just override this initial state with whatever was persisted.
`autoMergeLevel2` overcomes this problem, because it would first make a copy of the initial state, and then only override the keys within the auth object that were persisted. Since 'error' wouldn't be persisted, it would be left alone.

**Comment by rvirani1** [comment]
Just a note, I've had two developers on my team struggle with implementing v5. It was only after we found this PR did we really get a sense of how v5 worked. The current docs on the main README are still inadequate in our opinion. Would love for a version of this PR to get merged in.

**Comment by thenewt15** [comment]
Thanks @rvirani1 I may move this to a blog post if it remains here getting no love. There's actually a lot I'd like to clean up with the docs, but I prefer to see outstanding PRs merged in before I submit new ones.

**Comment by rvirani1** [comment]
A blog post would be great.

**Comment by ScreamZ** [comment]
Nice explanations, really loved that.

It could go a little deep in a concept though, like the **key** that can be used to make a "Vary" on userId per example.

Also detailing each config props a bit further but also especially the api https://github.com/rt2zz/redux-persist/blob/master/docs/api.md : stateReconciler, throttle, keyPrefix, ...

Thanks anyway.

Spread the love <3

**Comment by thenewt15** [comment]
@rvirani1 @ScreamZ I put this on medium, with an additional writeup on the merge process. Still hoping this gets merged at some point, but in the meantime I think this will help some people.

https://medium.com/@thenewt15/the-definitive-guide-to-redux-persist-84738167975

**Comment by rt2zz** [comment]
going to get this in now, will do a second revision pass after

**Comment by rt2zz** [comment]
thanks @thenewt15 !
